src/ansiblelint/utils.py

```diff
@@ -901,3 +909,80 @@ def nested_items(
+
+    This is a recursive function that calls itself for each nested layer of data.
+    Each iteration yields:
+
+    1. the current item's dictionary key or list index,
+    2. the current item's value, and
+    3. the path to the current item from the outermost data structure.
+
+    For dicts, the yielded (1) key and (2) value are what mydict.items() yields.
+    For lists, the yielded (1) index and (2) value are what enumerate(mylist) yields.
+    The final component, the parent path, is a list of dict keys and list indexes.
+    The parent path can be helpful in providing error messages that indicate
+    precisely which part of a yaml file (or other data strucutre) needs to be fixed.
+
+    For example, given this playbook:
+
+    .. code-block:: yaml
+
+        - name: a play
+          tasks:
+          - name: a task
+            debug:
+              msg: foobar
+
+    Here's the first and last yielded items:
+
+    .. code-block:: python
+
+        >>> playbook=[{"name": "a play", "tasks": [{"name": "a task", "debug": {"msg": "foobar"}}]}]
+        >>> next( nested_items_path( playbook ) )
+        (0, {'name': 'a play', 'tasks': [{'name': 'a task', 'debug': {'msg': 'foobar'}}]}, [])
+        >>> list( nested_items_path( playbook ) )[-1]
+        ('msg', 'foobar', [0, 'tasks', 0, 'debug'])
+
+    Note that, for outermost data structure, the parent path is ``[]`` because
+    you do not need to descend into any nested dicts or lists to find the indicated
+    key and value.
+
+    If a rule were designed to prohimit "foobar" debug messages, it could use the
+    parent path to provide a path to the problematic ``msg``. It might use a jq-style
+    path in its error message: "the error is at ``.[0].tasks[0].debug.msg``".
+    Or if a utility could automatically fix issues, it could use the path to descend
+    to the parent object using something like this:
+
+    .. code-block:: python
+
+        target = data
+        for segment in parent_path:
+            target = target[segment]
+
+    :param data: The nested data (dicts or lists).
+    :param parent_path: Do not use this param. It is used internally to recursively
+                        build the yielded parent path (list of keys, indexes).
+
+    :returns: each iteration yields the key (of the parent dict) or the index (lists)
+    """
+    if parent_path is None:
+        parent_path = []
+    if isinstance(data, dict):
+        convert_to_tuple = data.__class__.items  # subclasses of dict can override items
+    elif isinstance(data, list):
+        convert_to_tuple = enumerate
+    else:
+        raise TypeError(
+            f"nested_items_path expects a dict or a list but got {type(data)}"
+        )
+    for key, value in convert_to_tuple(data):
+        yield key, value, parent_path
+        if isinstance(value, (dict, list)):
+            yield from nested_items_path(value, parent_path + [key])
```

Comment:
> The only thing I don't understand is that why the second loop inverts key and value — I suppose it's a bug.

@webknjaz You said this might be a bug - but it is not a bug.
This recursive call is passing in `data=value` setting a new parent_path for the children of this `value`. That is how it creates the parent_path for each level of the nested data structure.

Maybe it would help to see just the recursive call lines in what I originally had (before switching to `yield from` and de-duplicating the loops):
```python
            for k, v, p in nested_items_path(value, parent_path + [key]):
            for k, v, p in nested_items_path(item, parent_path + [index]):
```
value is like item
key is like index

In any case, hopefully it's much clearer without the two separate loops.

Reply:
Interesting. I guess we need good test coverage for this since it's untrivial to follow and the tests would act as a demonstration of the usage/invocations + expected result example.